Libraries/LibMedia/Audio/PlaybackStreamWasapi.cpp

```diff
@@ -0,0 +1,512 @@
+    HANDLE task_event;
+
+    bool playing = false;
+    bool drain_and_suspend = false;
+    Atomic<bool> exit_requested = false;
+
+    static int render_thread_loop(AudioState& state);
+    RefPtr<Threading::Thread> audio_thread;
+    RefPtr<Core::ThreadedPromise<AK::Duration>> resume_promise;
+    RefPtr<Core::ThreadedPromise<void>> suspend_promise;
+
+    // I doubt we will have anyone with more than 7.1
+    Vector<float, ChannelMap::capacity()> channel_volumes;
+    UINT64 audio_client_clock_frequency;
+};
+
+PlaybackStreamWASAPI::AudioState::AudioState()
+{
+    task_event = CreateEvent(NULL, FALSE, FALSE, NULL);
+    VERIFY(task_event);
+}
+
+PlaybackStreamWASAPI::AudioState::~AudioState()
+{
+    CloseHandle(buffer_event);
+}
+
+ALWAYS_INLINE AK::Duration PlaybackStreamWASAPI::total_time_played_with_com_initialized(PlaybackStreamWASAPI::AudioState& state)
+{
+    UINT64 position;
+    MUST_HR(state.clock->GetPosition(&position, nullptr));
+    // FIXME: This is a narrowing conversion. We might want a u64 overload of from_nanoseconds.
+    return AK::Duration::from_time_units(AK::clamp_to<i64>(position), 1, state.audio_client_clock_frequency);
+}
+
+PlaybackStreamWASAPI::PlaybackStreamWASAPI(NonnullRefPtr<AudioState> state)
+    : m_state(move(state))
+{
+}
+
+PlaybackStreamWASAPI::~PlaybackStreamWASAPI()
+{
+    m_state->exit_requested.store(true, MemoryOrder::memory_order_release);
+    // Poke the event to wake the thread up from wait
+    VERIFY(m_state->buffer_event != NULL);
+    SetEvent(m_state->buffer_event);
+}
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStream::create(OutputState initial_output_state, u32 target_latency_ms, SampleSpecificationCallback&& specification_callback, AudioDataRequestCallback&& data_callback)
+{
+    return PlaybackStreamWASAPI::create(initial_output_state, target_latency_ms, move(specification_callback), move(data_callback));
+}
+
+static void print_audio_format(WAVEFORMATEXTENSIBLE& format)
+{
+    dbgln("wFormatTag: {:x}\n"
+          "nChannels: {}\n"
+          "nSamplesPerSec: {}\n"
+          "nAvgBytesPerSec: {}\n"
+          "nBlockAlign: {}\n"
+          "wBitsPerSample: {}\n"
+          "cbSize: {}\n"
+          "Samples.wValidBitsPerSample: {}\n"
+          "dwChannelMask: {:b}\n"
+          "SubFormat: {}",
+        format.Format.wFormatTag,
+        format.Format.nChannels,
+        format.Format.nSamplesPerSec,
+        format.Format.nAvgBytesPerSec,
+        format.Format.nBlockAlign,
+        format.Format.wBitsPerSample,
+        format.Format.cbSize,
+        format.Samples.wValidBitsPerSample,
+        format.dwChannelMask,
+        Span<u8> { reinterpret_cast<char*>(&format.SubFormat), 16 });
+}
+
+static ChannelMap convert_bitmask_to_channel_map(u32 channel_mask, u32 channels)
+{
+    Vector<Channel, ChannelMap::capacity()> channel_vec;
+    for (u64 i = 0; i < channels; i++) {
+        if (channel_mask & (1 << i))
```

Comment:
Doesn't this assume that the channel mask has bits 0 to `channels - 1` set, rather than checking all bits?